src/search/merge_and_shrink/labels.h

```diff
@@ -4,45 +4,65 @@
 #include <memory>
 #include <vector>
 
-namespace utils {
-class LogProxy;
-}
-
 namespace merge_and_shrink {
-class Label {
-    /*
-      This class implements labels as used by merge-and-shrink transition systems.
-      Labels are opaque tokens that have an associated cost.
-    */
-    int cost;
+class LabelsConstIterator {
+    const std::vector<int> &label_costs;
+    std::size_t current_index;
+
+    void next_valid_index();
 public:
-    explicit Label(int cost_)
-        : cost(cost_) {
+    LabelsConstIterator(const std::vector<int> &label_costs, bool end);
+    void operator++();
+
+    int operator*() const {
+        return int(current_index);
     }
-    ~Label() {}
-    int get_cost() const {
-        return cost;
+
+    bool operator==(const LabelsConstIterator &rhs) const {
+        return current_index == rhs.current_index;
+    }
+
+    bool operator!=(const LabelsConstIterator &rhs) const {
+        return current_index != rhs.current_index;
     }
 };
 
 /*
-  This class serves both as a container class to handle the set of all labels
-  and to perform label reduction on this set.
+  This class serves both as a container class to handle the set of all
+  labels and to perform label reduction on this set.
+
+  Labels are identified via integers indexing label_costs, which store their
+  costs. When using label reductions, reduced labels are set to -1 in label
+  costs.
 */
 class Labels {
-    std::vector<std::unique_ptr<Label>> labels;
-    int max_size; // the maximum number of labels that can be created
+    std::vector<int> label_costs;
+    int max_num_labels; // The maximum number of labels that can be created.
+    int num_active_labels; // The current number of active (non-reduced) labels.
 public:
-    explicit Labels(std::vector<std::unique_ptr<Label>> &&labels);
-    void reduce_labels(const std::vector<int> &old_label_nos);
-    bool is_current_label(int label_no) const;
-    int get_label_cost(int label_no) const;
-    void dump_labels(utils::LogProxy &log) const;
+    Labels(std::vector<int> &&label_costs, int max_num_labels);
+    void reduce_labels(const std::vector<int> &old_labels);
+    int get_label_cost(int label) const;
+    void dump_labels() const;
+
     int get_size() const {
-        return labels.size();
+        return label_costs.size();
```

Comment:
It's maybe not obvious what `get_size()` is there for/what the return value means; does it need an explanation? (Intuitively, it could mean either what it means here or the number of active labels.)

Reply:
Perhaps avoid the method name altogether and use one that resonates with `get_num_active_labels`, e.g. using `total` or whatever. But perhaps it doesn't need changing.